Create bisq easy trade details view

makhni-dev opened the pull request:
Closes bisq-network/bisq2#2568
This commit creates a pop up screen that contains trade details of a bisq easy trade. The design is similar to that used in trade wizard review offer page.
The open button looks like this: 
![Screenshot from 2024-10-25 14-02-21](https://github.com/user-attachments/assets/db9da8de-339c-4879-9be1-e366b4959baa)
This opens the following popup:
<img src="https://github.com/user-attachments/assets/306bdd3f-f62b-44a5-b129-795d629a909c" alt="Trade details" width="600" />

**Benefits**
This view can act as a central point for displaying important trade data, which is currently displayed in several separate locations, meaning this might increase clarity for users. It currently does not display anything that is not already visible in some other way, but this can easily be extended if wanted. 

**Future work**
Several more fields can be added here that can be of use, for example:
-  time since last user activity from peer
-  transaction ID
-  blockchain confirmation
**I suggest that I write an issue where we all can input what more fields that would be nice to have when this is merged.**

HenrikJannsen commented:
Thanks for the PR! Here some comments.
In minimal app size the popup's height should not exceed the app height.
![Screenshot 2024-10-26 at 13 02 19](https://github.com/user-attachments/assets/6b3116cc-bf0b-4e61-84e0-a9316c55efb7)

I understand that you wanted to avoid a vertical scrollbar, which I agree is preferred if possible, but here I think we either minimize further the spacing of the rows or accept scrolling. We have usually design with lot of spacing, so I think a scrollbar here and following design from other screens is better here. Maybe @axpoems can add his input here or suggest a design?

My role: missing buyer/seller role
Market price should be Trade price, we can/should add the information if market price, float price or fixed price was used.
Amount in Fiat: Better use as well a textfield to allow copy

Fields which have no provided data yet: we could also just hide them. Or maybe better add another section of process-related data (e.g. data which will be set during the trade process, there will be more like blockchain confirmation, tx id,...)
In that case we should show all fields.

My username/My network address: I think we can drop those
Offer taken time stamp: We use the term date -> Trade date

Further important data:
Selected mediator (can be empty but is not in live network)

You could use instead of text fields the `CopyOnClickLabel`. Then the layout could be a bit condensed.

HenrikJannsen commented on apps/desktop/desktop/src/main/java/bisq/desktop/main/content/bisq_easy/open_trades/trade_details/TradeDetailsController.java:
```diff
@@ -0,0 +1,150 @@
+    public void initWithData(InitData initData) {
+        BisqEasyTrade trade = initData.getBisqEasyTrade();
+        BisqEasyOpenTradeChannel channel = initData.channel;
+        model.setTradeId(new SimpleStringProperty(trade.getId()));
```
We prefer to use final fields in the model for properties. Not a big problem to do it like that but better to stick with the established style.

HenrikJannsen commented on apps/desktop/desktop/src/main/java/bisq/desktop/main/content/bisq_easy/open_trades/trade_details/TradeDetailsController.java:
```diff
@@ -0,0 +1,150 @@
+        String myAddress = formatNetworkAddresses(myNetworkId.getAddressByTransportTypeMap());
+        model.setMyNetworkAddress(new SimpleStringProperty(myAddress));
+        
+        String amountInFiat = AmountFormatter.formatAmount(quoteAmount);
```
Better use `AmountFormatter.formatAmountWithCode`

HenrikJannsen commented on apps/desktop/desktop/src/main/java/bisq/desktop/main/content/bisq_easy/open_trades/trade_details/TradeDetailsController.java:
```diff
@@ -0,0 +1,150 @@
+
+    public String formatNetworkAddresses(AddressByTransportTypeMap addressMap) {
+        return Joiner.on("; ").join(addressMap.entrySet().stream()
+                .map(e -> StringUtils.truncate(e.getValue().getFullAddress(), Integer.MAX_VALUE))
```
Why use truncate when setting Integer.MAX_VALUE?

HenrikJannsen commented on apps/desktop/desktop/src/main/java/bisq/desktop/main/content/bisq_easy/open_trades/trade_details/TradeDetailsModel.java:
```diff
@@ -0,0 +1,66 @@
+        return NavigationTarget.BISQ_EASY_TRADE_DETAILS;
+    }
+
+    public SimpleStringProperty getBitcoinPaymentAddress() {
```
We use models as pure value objects, any logic even simple should be done in the controller.

HenrikJannsen replied:
In that case I would just use the dataNotYetProvided as the default value in the property and set the property with a value once its gets set.

HenrikJannsen commented on apps/desktop/desktop/src/main/java/bisq/desktop/main/content/bisq_easy/open_trades/trade_details/TradeDetailsView.java:
```diff
@@ -0,0 +1,223 @@
+    private final Button closeButton;
+    private final ScrollPane scrollPane;
+
+    private final VBox allInfoContainer, informationContainer, detailsContainer;
```
Can be local as well as scrollPane. IntelliJ shows warnings for such issues.

HenrikJannsen commented on apps/desktop/desktop/src/main/java/bisq/desktop/main/content/bisq_easy/open_trades/trade_details/TradeDetailsView.java:
```diff
@@ -0,0 +1,223 @@
+        scrollPane.setFitToWidth(true);
+        scrollPane.setPadding(new Insets(15, 15, 15, 15));
+        allInfoContainer.setPadding(new Insets(30, 30, 30, 30));
+        allInfoContainer.prefHeightProperty().bind(root.heightProperty());
```
bindings must be set in onViewAttached. Also follow established style to use onViewAttached and onViewDetached after constructor as those are the most important methods.

All bindings, event listeners, event handlers must be deactivated in onViewDetached. See the dev docs as well or other code.

HenrikJannsen commented on apps/desktop/desktop/src/main/java/bisq/desktop/main/content/bisq_easy/open_trades/trade_details/TradeDetailsView.java:
```diff
@@ -0,0 +1,223 @@
+    }
+
+    private void setupColorChangeEffects() {
+        bitcoinPaymentAddress.textField.textProperty().addListener(new ChangeListener<String>() {
```
listeners need to be stored in a field and removed at onViewDetached.

HenrikJannsen commented on apps/desktop/desktop/src/main/java/bisq/desktop/main/content/bisq_easy/open_trades/trade_state/TradeStateView.java:
```diff
@@ -186,6 +194,7 @@ protected void onViewAttached() {
         });
 
         cancelButton.setOnAction(e -> controller.onInterruptTrade());
+        tradeDetailsButton.setOnAction(e -> controller.onViewTradeDetails());
```
Missing the `tradeDetailsButton.setOnAction(null)` in onViewDetached

makhni-dev commented:
@HenrikJannsen Thank you for the review! I will work on the requested changes tomorrow and if I dont have time to finish, continue on friday.

makhni-dev commented:
@HenrikJannsen I have fixed the majority of the requests on my computer but have two questions:
> Fields which have no provided data yet: we could also just hide them. Or maybe better add another section of process-related data (e.g. data which will be set during the trade process, there will be more like blockchain confirmation, tx id,...) In that case we should show all fields.

This is a great idea, do you want the blockchain confirmation and tx id to be filled in in this pull request or should it be done separately, possibly in another issue?



> Further important data: Selected mediator (can be empty but is not in live network)

I have searched but not found a good way to test mediator assignment on localhost, is there some way to assign a mediator through config or JVM options somehow?

makhni-dev commented:
@axpoems and @HenrikJannsen, I have now updated the design to be much more in line with that used in TradeWizard review offer page. Please have a look and tell me what you think! I have updated the pull request description as well.

HenrikJannsen reviewed (approved):
utACK
